tests/suites/test_suite_psa_crypto_driver_wrappers.function

```diff
@@ -1784,3 +1784,126 @@ exit:
     mbedtls_test_driver_hash_hooks = mbedtls_test_driver_hash_hooks_init();
 }
 /* END_CASE */
+
+/* BEGIN_CASE */
+void asymmetric_encrypt_decrypt( data_t *key_data,
+                                 data_t *input_data,
+                                 data_t *label,
```

Comment:
I'm not sure if such coupling will work for every case. The way it works now, we won't be able to test decryption if encryption doesn't succeed. Since the output is partially random, we might assume that for real implementations it's not possible to decouple these operations, but for fake drivers and forced outputs - we can test them separately.

Reply:
Indeed we need to be able to test encryption and decryption separately. Otherwise we could miss bugs that cause an encrypt-decrypt sequence to work only if the encryption step left some data around in an improperly cleared data structure.

Reply:
So what solution would you suggest, @gilles-peskine-arm?
Will setting a dummy entropy source make the output deterministic, thus, letting us test encryption and decryption independently?
Or maybe these tests should stay as they are, and just two additional ones should be added with a dummy entropy and minimal sanity cases, independently for encryption and decryption?

Reply:
I would appropriate any hints regarding the testing approach. I picked enctypt/decrypt approach as only this worked for me and I didn't know how to test encryption and decryption separately.

Reply:
I did some investigation around encryption and `PSA_ALG_RSA_PKCS1V15_CRYPT` and `PSA_ALG_RSA_OAEP` algs. 

Encryption for both algs gives different output for each call. Does it mean that both are not deterministic? Results below:
```
przemek@przemek-VirtualBox:~/src/mbedtls/tests$ ./test_suite_psa_crypto_driver_wrappers
PSA encrypt transparent driver: in-driver RSA PKCS#1 v1.5 .........  
 encrypt out: 7db488b97b1b436aaf06ad600ea31cfb8b8058dbc8a2ef755ae5f2c16b0f9f0f55c7d21ab248faad39cf7d8dcc7133db7ad257c289a114b0d280a4e98b227ab0a18de85a92cd5fb04f8087803ccf131b044d2f12f2337b7300e6258eb23205cf4d47304a9e0c8054178cf858fe31ff39949c88e873a07335ac38431c4b52ea3d 
 length out: 128
PASS
PSA encrypt transparent driver: in-driver RSA OAEP-SHA-256 ........  
 encrypt out: a9fd11426b1305a5b709398cb0b8797da42f8eee45fa740cf91dd7bd6f717bee266c850e200b9383080dd881ef2e80aea10288cd8dd799da70371181551cd9fa263a645deecfc9b48dbd165380fbd9aca0b6b856cd6217f6f1f2dbef48317637dba62f8ff10fc8bbaaf9b489e3d46737422cf5e1b0e82477674e4fa88571bc7e 
 length out: 128
PASS

----------------------------------------------------------------------------

PASSED (2 / 2 tests (0 skipped))
przemek@przemek-VirtualBox:~/src/mbedtls/tests$ ./test_suite_psa_crypto_driver_wrappers
PSA encrypt transparent driver: in-driver RSA PKCS#1 v1.5 .........  
 encrypt out: 02dd353f533e8c4539686c6fdbf653028c1a39cdf2f5ac107f3ce040ac3e5ffddb19e59a76672c81229b9ed940ee78e7146efe837bbadf34ee980d7df78034e1bd08047fe34c5e84447d1a8b1846978f37dd905c7ea3fab4ace09324b2b1295157655bf3622f4c1869323aa645051db938fed8d450b105487072fc093f66e1bb 
 length out: 128
PASS
PSA encrypt transparent driver: in-driver RSA OAEP-SHA-256 ........  
 encrypt out: 197e725b05e36370a9707772f44fdd2783de676e0485fcead60b18561762993f18a16398db0082f2954a208b2d451c7debb84642453b26f9639df47e1d03617344e46a9b53bb9f4c390b179f04259c8c5f02d5599e590651d6b351054296b6b188ed88c5bb174a55df51b069c8dba50d95c8679d9f1aa2e36574753eefb09c3b 
 length out: 128
PASS

```

Then I tried to use fake/static seed based on the following test:
https://github.com/ARMmbed/mbedtls/blob/6d2479516c8373ed8a335f664ee4e05c4d12dd62/tests/suites/test_suite_psa_crypto_init.function#L263-L278

With the static seed got the same encryption output in successive runs for both algs:
```
przemek@przemek-VirtualBox:~/src/mbedtls/tests$ ./test_suite_psa_crypto_driver_wrappers
PSA encrypt transparent driver: in-driver RSA PKCS#1 v1.5 .........  
 encrypt out: 2dfabbd0dd2cd9bd61f122615dae67d3952f6b6c169d924ae0ca452eb71bc2075875039b0bcf1781bbb42887e1d26b7727a294050ca3df9ea5b44c4b3052164f66a1ff8723393d64fb76afe86fc3dae97b1a86ddde6dd1aa697f20d81b5c559780c8f5c6b2919676203a52de10a5f6e76ac218642072f4868085555345f26a61 
 length out: 128
PASS
PSA encrypt transparent driver: in-driver RSA OAEP-SHA-256 ........  
 encrypt out: 13fd8220c266b34b894f916448d854f7438e075f55cc4df95edf7002bd88fd4b156305112f48362c1563a490ae56c29b546edb31dd5901db0c60197724d0a56ba1a33bfb0ed23bc806b0b2ca87ae9323ff86c7a06bf05108281a324b9f9af0bd50220aaf003606be1e5333fb53b97dff13261f8c6d2e4a5cefcd2b5589eb2a21 
 length out: 128
PASS

----------------------------------------------------------------------------

PASSED (2 / 2 tests (0 skipped))
przemek@przemek-VirtualBox:~/src/mbedtls/tests$ ./test_suite_psa_crypto_driver_wrappers
PSA encrypt transparent driver: in-driver RSA PKCS#1 v1.5 .........  
 encrypt out: 2dfabbd0dd2cd9bd61f122615dae67d3952f6b6c169d924ae0ca452eb71bc2075875039b0bcf1781bbb42887e1d26b7727a294050ca3df9ea5b44c4b3052164f66a1ff8723393d64fb76afe86fc3dae97b1a86ddde6dd1aa697f20d81b5c559780c8f5c6b2919676203a52de10a5f6e76ac218642072f4868085555345f26a61 
 length out: 128
PASS
PSA encrypt transparent driver: in-driver RSA OAEP-SHA-256 ........  
 encrypt out: 13fd8220c266b34b894f916448d854f7438e075f55cc4df95edf7002bd88fd4b156305112f48362c1563a490ae56c29b546edb31dd5901db0c60197724d0a56ba1a33bfb0ed23bc806b0b2ca87ae9323ff86c7a06bf05108281a324b9f9af0bd50220aaf003606be1e5333fb53b97dff13261f8c6d2e4a5cefcd2b5589eb2a21 
 length out: 128
PASS

----------------------------------------------------------------------------

```

> We can have known-answer tests for deterministic algorithms. PSA_ALG_RSA_PKCS1V15_CRYPT is deterministic, and PSA_ALG_RSA_OAEP decryption as well. PSA_ALG_RSA_OAEP encryption is not determinstic.

I don't understand how one alg is deterministic and the other is not (encryption case)? I'm getting consistent results for encryption in both cases (normal: different result for both, fake seed: same result for both).

I also couldn't find known-answer test vectors for encryption and decryption. Seems like in current implementation output is not verified in `asymmetric_encrypt` psa test:
https://github.com/ARMmbed/mbedtls/blob/6d2479516c8373ed8a335f664ee4e05c4d12dd62/tests/suites/test_suite_psa_crypto.function#L5890-L5962

Reply:
Oops. Sorry. What I wrote above was stupid. No asymmetric _encryption_ method can be deterministic (signature can) — with deterministic asymmetric encryption, anyone who wants to decrypt a partially-known ciphertext could try to guess the plaintext, encrypt it (this only requires the public key) and compare with the ciphertext.

So unfortunately I think the best we can realistically do for now is known-answer tests for decryption, and an encrypt-decrypt cycle for encryption.

Later I'd like to have encrypt-decrypt tests where encryption and decryption are done by different implementation (built-in against driver). But that requires quite a bit more work on the testing architecture first.

Reply:
Added known-answer tests for decryption.